src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/SignerInfo.cs

```diff
@@ -120,18 +120,94 @@ public SignerInfoCollection CounterSignerInfos
 
         public Oid SignatureAlgorithm => new Oid(_signatureAlgorithm);
 
-        public void AddUnsignedAttribute(AsnEncodedData unsignedAttribute)
+        private delegate void WithSelfInfoDelegate(ref SignerInfoAsn mySigned);
+
+        private void WithSelfInfo(WithSelfInfoDelegate action)
         {
-            int myIdx = _document.SignerInfos.FindIndexForSigner(this);
+            if (_parentSignerInfo == null)
+            {
+                int myIdx = _document.SignerInfos.FindIndexForSigner(this);
 
-            if (myIdx < 0)
+                if (myIdx < 0)
+                {
+                    throw new CryptographicException(SR.Cryptography_Cms_SignerNotFound);
+                }
+
+                ref SignedDataAsn signedData = ref _document.GetRawData();
+                ref SignerInfoAsn mySigner = ref signedData.SignerInfos[myIdx];
+
+                action(ref mySigner);
+
+                // Re-normalize the document
+                _document.Reencode();
+            }
+            else
             {
-                throw new CryptographicException(SR.Cryptography_Cms_SignerNotFound);
+                // we are one level deep, we need to update signer and counter signer attributes
+                int parentIdx = _document.SignerInfos.FindIndexForSigner(_parentSignerInfo);
+
+                if (parentIdx == -1)
+                {
+                    throw new CryptographicException(SR.Cryptography_Cms_NoSignerAtIndex);
```

Comment:
Did this one get a test?

Reply:
yes (it has `_Throws` suffix - one of them tests this scenario)